File: Domain/Head.cs

using System.Runtime.Serialization;

[DataContract]
public class Head<T> : IDocumentKey where T : IDocumentKey
{
    [DataMember(Name = "id")]
    public string Id { get; set; }

    [DataMember(Name = "content")]
    public T Content { get; set; }

    [DataMember(Name = "version_item_id")]
    public string VersionItemId { get; set; }

    public string Key => $"{Type}::{Id}";

    [DataMember(Name = "type")]
    public string Type { get; } = "head";
}

File: DatabaseTypes/Head.fs
namespace DatabaseTypes

open System.Text.Json.Serialization
open DatabaseTypes.Identificators
open Utils.Converters

[<JsonConverter(typeof<HeadTypeConverter>)>]
type HeadType private () =
    inherit BaseType("head")
    static member Instance = HeadType()
and HeadTypeConverter() =
    inherit StringConverter<HeadType>((fun m -> m.Value), (fun _ -> HeadType.Instance))

type Head =
    {
        [<JsonPropertyName("id")>]
        Id : HeadId
        [<JsonPropertyName("name")>]
        Name : string
        [<JsonPropertyName("commit")>]
        Commit : Commit
        [<JsonPropertyName("permissions")>]
        Permissions : Permissions
        [<JsonPropertyName("type")>]
        Type : HeadType
    }

    static member CreateDocumentKey(id: HeadId): DocumentKey =
        DocumentKey.Create(id.Value, HeadType.Instance.Value)
    member private this.DocKey = Head.CreateDocumentKey(this.Id)

    interface IDocumentKey with
        member this.Type
            with get() = this.DocKey.Type
        member this.Key
            with get() = this.DocKey.Key


